champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/Skill.java

```diff
@@ -392,4 +406,26 @@ protected int getLevel(Player player) {
     public boolean isHolding(Player player) {
         return hasSkill(player) && SkillWeapons.isHolding(player, getType());
     }
+
+    public static <T extends Projectile> void updateParticleForArrowTrail(Function<Location, ParticleBuilder> builderFunction,
+                                                                          Iterator<T> it, boolean checkIfInBlock) {
+        updateParticleForArrowTrail(builderFunction, it, checkIfInBlock, new Vector(0, 0, 0));
+    }
+
+    public static <T extends Projectile> void updateParticleForArrowTrail(Function<Location, ParticleBuilder> builderFunction,
+                                                                          Iterator<T> it, boolean checkIfInBlock,
+                                                                          Vector vector) {
+        while (it.hasNext()) {
+            T arrowOrProjectile = it.next();
+            if (arrowOrProjectile == null || arrowOrProjectile.isDead()) {
+                it.remove();
+            } else if (checkIfInBlock && arrowOrProjectile instanceof Arrow && ((Arrow) arrowOrProjectile).isInBlock()) {
+                it.remove();
+            } else {
+                Location location = arrowOrProjectile.getLocation().add(vector);
+                ParticleBuilder particleBuilder = builderFunction.apply(location);
+                particleBuilder.spawn();
+            }
+        }
+    }
```

Comment:
Feel like you could move this to the PrepareArrowSkill subclass and then handle the update events for all of them within that as well